src/Console/TestSuite/ConsoleIntegrationTestTrait.php

```diff
@@ -0,0 +1,345 @@
+ * @since         3.7.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Console\TestSuite;
+
+use Cake\Command\Command;
+use Cake\Console\CommandRunner;
+use Cake\Console\ConsoleIo;
+use Cake\Console\Exception\StopException;
+use Cake\Console\TestSuite\Constraint\ContentsContain;
+use Cake\Console\TestSuite\Constraint\ContentsContainRow;
+use Cake\Console\TestSuite\Constraint\ContentsEmpty;
+use Cake\Console\TestSuite\Constraint\ContentsNotContain;
+use Cake\Console\TestSuite\Constraint\ContentsRegExp;
+use Cake\Console\TestSuite\Constraint\ExitCode;
+use Cake\Core\TestSuite\ContainerStubTrait;
+use RuntimeException;
+
+/**
+ * A bundle of methods that makes testing commands
+ * and shell classes easier.
+ *
+ * Enables you to call commands/shells with a
+ * full application context.
+ */
+trait ConsoleIntegrationTestTrait
+{
+    use ContainerStubTrait;
+
+    /**
+     * Whether to use the CommandRunner
+     *
+     * @var bool
+     */
+    protected $_useCommandRunner = false;
+
+    /**
+     * Last exit code
+     *
+     * @var int|null
+     */
+    protected $_exitCode;
+
+    /**
+     * Console output stub
+     *
+     * @var \Cake\Console\TestSuite\StubConsoleOutput
+     */
+    protected $_out;
+
+    /**
+     * Console error output stub
+     *
+     * @var \Cake\Console\TestSuite\StubConsoleOutput
+     */
+    protected $_err;
+
+    /**
+     * Console input mock
+     *
+     * @var \Cake\Console\TestSuite\StubConsoleInput
+     */
+    protected $_in;
+
+    /**
+     * Runs CLI integration test
+     *
+     * @param string $command Command to run
+     * @param array $input Input values to pass to an interactive shell
+     * @throws \Cake\Console\TestSuite\MissingConsoleInputException
+     * @throws \RuntimeException
+     * @return void
+     */
+    public function exec(string $command, array $input = []): void
+    {
+        $runner = $this->makeRunner();
+
+        if ($this->_out === null) {
+            $this->_out = new StubConsoleOutput();
+        }
+        if ($this->_err === null) {
+            $this->_err = new StubConsoleOutput();
+        }
+        if ($this->_in === null) {
+            $this->_in = new StubConsoleInput($input);
+        } elseif ($input) {
+            throw new RuntimeException('You can use `$input` only if `$_in` property is null and will be reset.');
+        }
+
+        $args = $this->commandStringToArgs("cake $command");
+        $io = new ConsoleIo($this->_out, $this->_err, $this->_in);
+
+        try {
+            $this->_exitCode = $runner->run($args, $io);
+        } catch (MissingConsoleInputException $e) {
+            $messages = $this->_out->messages();
+            if (count($messages)) {
+                $e->setQuestion($messages[count($messages) - 1]);
+            }
+            throw $e;
+        } catch (StopException $exception) {
+            $this->_exitCode = $exception->getCode();
```

Comment:
@markstory should we be testing this?

Reply:
Probably. I thought we would have had a test already 🤔